src/rust/engine/process_executor/src/test_server.rs

```diff
@@ -0,0 +1,159 @@
+use std::fmt::Debug;
+use std::ops::Deref;
+use std::slice::Iter;
+use std::sync::{Arc, Mutex};
+
+use bazel_protos;
+use grpcio;
+use protobuf;
+
+#[derive(Clone, Debug)]
+pub struct MockExecution {
+  ///
+  /// The name of the operation. It is assumed that all operation_responses use this name.
+  ///
+  pub name: String,
+  ///
+  /// The expected ExecuteRequest.
+  ///
+  pub execute_request: bazel_protos::remote_execution::ExecuteRequest,
+  ///
+  /// Vec of Operation response for Execution or GetOperation requests. Will be returned in order.
+  ///
+  pub operation_responses: Vec<bazel_protos::operations::Operation>,
+}
+
+///
+/// A server which will answer ExecuteRequest and GetOperation gRPC requests with pre-canned responses.
+///
+pub struct TestServer {
+  mock_server: MockServer,
+  server_transport: grpcio::Server,
+}
+
+impl TestServer {
+  ///
+  /// # Arguments
+  /// * `mock_execution` - The canned responses to issue. Returns the MockExecution's
+  ///                      operation_responses in order to any ExecuteRequest or GetOperation
+  ///                      requests.
+  ///
+  ///                      If an ExecuteRequest request is received which is not equal to this
+  ///                      MockExecution's execute_request, an error will be returned.
+  ///
+  ///                      If a GetOperation request is received whose name is not equal to this
+  ///                      MockExecution's name, or more requests are received than stub responses
+  ///                      are available for, an error will be returned.
+  pub fn new(mock_execution: MockExecution) -> TestServer {
+    let mock_server = MockServer::new(mock_execution);
+
+    let env = Arc::new(grpcio::Environment::new(1));
+    let mut server_transport = grpcio::ServerBuilder::new(env)
+        .register_service(bazel_protos::remote_execution_grpc::create_execution(mock_server.clone()))
+        .register_service(bazel_protos::operations_grpc::create_operations(mock_server.clone()))
+        .bind("localhost", 0)
+        .build()
+        .unwrap();
+    server_transport.start();
+
+    TestServer {
+      mock_server,
+      server_transport,
+    }
+  }
+
+  ///
+  /// The address on which this server is listening over insecure HTTP transport.
+  ///
+  pub fn address(&self) -> String {
```

Comment:
Would it be overkill to use https://doc.rust-lang.org/std/net/enum.SocketAddr.html here to get some type safety?

Reply:
Right now I'm binding to localhost which is IPv4/6-agnostic, and to use a SocketAddr I would either need to pick an IP version, or resolve manually localhost, both of which seem worse than the status quo.